contracts/src/systems/combat/contracts/battle_systems.cairo

```diff
@@ -1155,6 +1183,15 @@ pub impl InternalBattleImpl of InternalBattleTrait {
 
     /// Make army leave battle
     fn leave_battle(ref world: WorldStorage, ref battle: Battle, ref original_army: Army) {
+        // [Achievement] Win a battle
+        if battle.has_ended() && original_army.troops.count() > 0 {
```

Comment:
you should make a variable at the top to know if battle ended, then use the updated army at the bottom when checking if army count is > 0